mmeval/metrics/rouge.py

```diff
@@ -0,0 +1,288 @@
+        elif isinstance(tokenizer, str):
+            self.tokenizer_fn = get_tokenizer(tokenizer)
+            if self.tokenizer_fn is None:
+                raise ValueError('Right now, `tokenizer` only supports '
+                                 "pre-defined 'en' or 'cn'.")
+        else:
+            assert tokenizer is None, \
+                f'`tokenizer` supports Callable, str or None, but not `{type(tokenizer)}`'  # noqa: E501
+        self.accumulate = accumulate
+
+    def add(self, predictions: Sequence[str], references: Sequence[Sequence[str]]) -> None:  # type: ignore # yapf: disable # noqa: E501
+        """Add the intermediate results to ``self._results``.
+
+        Args:
+             predictions (Sequence[str]): An iterable of predicted sentences.
+             references (Sequence[Sequence[str]): An iterable of
+                 referenced sentences.
+        """
+        # If the tokenizer is None, check the first sentence
+        # to determine which language the tokenizer is used.
+        if self.tokenizer_fn is None:
+            language = inference_language(predictions[0])
+            self.tokenizer_fn = get_tokenizer(language)
+        # Traverse the predicted sentences
+        for prediction, _references in zip(predictions, references):
+            pred_token = self._normalize_and_tokenize(prediction)
+            reference_len = len(_references)
+            ref_tokens = []
+            results = []
+            for reference in _references:
+                ref_token = self._normalize_and_tokenize(reference)
+                ref_tokens.append(ref_token)
+            # Traverse the chosen rouge_keys
+            for i, rouge_key in enumerate(self.rouge_keys):
+                fmeasure = precision = recall = 0.
+                # Traverse the tokens of references for single prediction.
+                for j, ref_token in enumerate(ref_tokens):
+                    if isinstance(rouge_key, int):
+                        score = _rougeN_score(pred_token, ref_token, rouge_key)
+                    else:
+                        score = _rougeL_score(pred_token, ref_token)
+                    if self.accumulate == 'best':
+                        if fmeasure < score['fmeasure']:
+                            fmeasure = score['fmeasure']
+                            precision = score['precision']
+                            recall = score['recall']
+                    else:
+                        fmeasure += score['fmeasure']
+                        precision += score['precision']
+                        recall += score['recall']
+                if self.accumulate == 'avg':
+                    fmeasure = fmeasure / reference_len
+                    precision = precision / reference_len
+                    recall = recall / reference_len
+
+                result = (fmeasure, precision, recall)
+
+                results.append(result)
+            self._results.append(results)
```

Comment:
Suggest move the logic of calculating the score into a function named `_compute_rouge_score`:

```python
def add(self, predictions: Sequence[str], references: Sequence[Sequence[str]]) -> None:  # type: ignore # yapf: disable # noqa: E501
        """Add the intermediate results to ``self._results``.

        Args:
             predictions (Sequence[str]): An iterable of predicted sentences.
             references (Sequence[Sequence[str]): An iterable of
                 referenced sentences.
        """
        # If the tokenizer is None, check the first sentence
        # to determine which language the tokenizer is used.
        if self.tokenizer_fn is None:
            language = inference_language(predictions[0])
            self.tokenizer_fn = get_tokenizer(language)

        # Traverse the predicted sentences
        for prediction, _references in zip(predictions, references):
            scores_per_rouge_keys = self._compute_rouge_score(prediction, _references)
            self._results.append(scores_per_rouge_keys)

    def _compute_rouge_score(self, prediction, refernces):
        pred_token = self._normalize_and_tokenize(prediction)
        ref_tokens = [self._normalize_and_tokenize(refs) for refs in refernces]

        # Traverse the chosen rouge_keys
        scores_per_rouge_keys = []
        for rouge_key in self.rouge_keys:

            # Traverse the tokens of references for single prediction.
            scores = []
            for ref_token in ref_tokens:
                if isinstance(rouge_key, int):
                    score = _rougeN_score(pred_token, ref_token, rouge_key)
                else:
                    score = _rougeL_score(pred_token, ref_token)
                scores.append(score)

            # # Accumulate rouge score across multiple refernce.
            if len(scores) == 0:
                fmeasure = precision = recall = 0.0
            if self.accumulate == 'best':
                fmeasure = max([score['fmeasure'] for score in scores])
                precision = max([score['precision'] for score in scores])
                recall = max([score['recall'] for score in scores])
            else:
                fmeasure = sum([score['fmeasure'] for score in scores]) / len(scores)
                precision = sum([score['precision'] for score in scores]) / len(scores)
                recall = sum([score['recall'] for score in scores]) / len(scores)

            scores_per_rouge_keys.append((fmeasure, precision, recall))

        return scores_per_rouge_keys
```